contracts/staking/EpochSelector.sol

```diff
@@ -0,0 +1,1705 @@
+    function search(address _node) public view returns (bool) {
+        if (_node == address(0)) {
+            return false;
+        }
+        Node memory node = nodes[_node];
+        return node.node == _node && node.balance != 0;
+    }
+
+    /// @notice Internal function to delete the node from the key
+    /// @param _root Current root
+    /// @param key Address of the node to be removed
+    /// @param existingBalanceOfKey Balance of the key to be deleted
+    function _deleteNode(
+        address _root,
+        address key,
+        uint96 existingBalanceOfKey
+    ) internal returns (address) {
+        // console2.log("At node", _root);
+        // console2.log("Element to delete", key);
+        // console2.log("Balance of key to delete", existingBalanceOfKey);
+        if (_root == address(0)) {
+            return (_root);
+        }
+
+        Node storage node = nodes[_root];
+        if (key < _root) {
+            // console2.log("Moving to left");
+            node.sumOfLeftBalances -= existingBalanceOfKey;
+            (node.left) = _deleteNode(node.left, key, existingBalanceOfKey);
+            // console2.log("After Moving to left");
+        } else if (key > _root) {
+            // console2.log("Moving to right");
+            // console2.log("node.sumOfRightBalances", node.sumOfRightBalances);
+            node.sumOfRightBalances -= existingBalanceOfKey;
+            (node.right) = _deleteNode(node.right, key, existingBalanceOfKey);
+            // console2.log("After Moving to right");
+        } else {
+            // console2.log("Wow! found node to delete");
+            // if node.left and node.right are full, select the next smallest element to node.right, replace it with element to be removed
+            // if node.right is full and node.left is null, select the next smallest element to node.right, replace it with element to be removed
+            // if node.left is full and node.right is null, select node.left, replace it with node.left
+            // if node.left and node.right are null, simply delete the element
+
+            if (node.left != address(0) && node.right != address(0)) {
+                // console2.log("case 1");
+                return _case1OnDelete(_root);
+            } else if (node.left == address(0) && node.right != address(0)) {
+                // console2.log("case 2");
+                return _case2OnDelete(_root);
+            } else if (node.left != address(0) && node.right == address(0)) {
+                // console2.log("case 3");
+                return _case3OnDelete(_root);
+            } else if (node.left == address(0) && node.right == address(0)) {
+                delete nodes[_root];
+                return address(0);
+            } else {
+                revert(ClusterLib.ERROR_OCCURED_DURING_DELETE);
+            }
+        }
+
+        node.height = calculateUpdatedHeight(node);
+
+        int256 balance = getHeightDifference(_root);
```

Comment:
Rebalance logic is reused in insert as well, so may be move this code to a different function.